File: crux-src/scanner.c
#include "scanner.h"

#include <stdbool.h>
#include <string.h>

typedef struct {
	const char *start;
	const char *current;
	int line;
} Scanner;

Scanner scanner;

/**
 * Advances the scanner to the next character.
 * @return The character that was just consumed
 */
static char advance(void)
{
	scanner.current++;
	return scanner.current[-1];
}

/**
 * Returns the current character without consuming it.
 * @return The current character
 */
static char peek(void)
{
	return *scanner.current;
}

/**
 * Checks if the scanner has reached the end of the source.
 * @return true if at the end of source, false otherwise
 */
static bool is_at_end(void)
{
	return *scanner.current == '\0';
}

/**
 * Returns the character after the current one without consuming any.
 * @return The next character, or '\0' if at the end of source
 */
static char peek_next(void)
{
	if (is_at_end())
		return '\0';
	return scanner.current[1];
}

/**
 * Determines if a character can start an identifier.
 * @param c The character to check
 * @return true if the character can start an identifier, false otherwise
 */
static bool is_identifier_starter(const char c)
{
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

/**
 * Checks if a character is alphabetic or underscore.
 * @param c The character to check
 * @return true if the character is alphabetic or underscore, false otherwise
 */
static bool is_alpha(const char c)
{
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

/**
 * Skips whitespace and comments in the source code.
 */
static void skip_whitespace(void)
{
	for (;;) {
		const char c = peek();
		switch (c) {
		case ' ':
		case '\r':
		case '\t':
			advance();
			break;
		case '\n': {
			scanner.line++;
			advance();
			break;
		}
		case '/': {
			if (peek_next() == '/') {
				// Comments are only on one line
				while (peek() != '\n' && !is_at_end())
					advance();
			} else {
				return;
			}
			break;
		}
		default:
			return;
		}
	}
}

static CruxTokenType check_keyword(const int start, const int length,
				   const char *rest, const CruxTokenType type)
{
	if (scanner.current - scanner.start == start + length &&
	    memcmp(scanner.start + start, rest, length) == 0) {
		return type;
	}
	return TOKEN_IDENTIFIER;
}

/**
 * Determines the token type of identifier.
 * @return The token type (keyword or identifier)
 */
static CruxTokenType identifier_type(void)
{
	switch (scanner.start[0]) {
	case 'a': {
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 's': {
				return check_keyword(2, 0, "", TOKEN_AS);
			}
			case 'n': {
				return check_keyword(2, 1, "d", TOKEN_AND);
			}
			default:;
			}
		}
		break;
	}
	case 'b':
		return check_keyword(1, 4, "reak", TOKEN_BREAK);
	case 'c': {
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 'o':
				return check_keyword(2, 6, "ntinue",
						     TOKEN_CONTINUE);
			default:;
			}
		}
		break;
	}
	case 'd': {
		return check_keyword(1, 6, "efault", TOKEN_DEFAULT);
	}
	case 'e':
		return check_keyword(1, 3, "lse", TOKEN_ELSE);
	case 'g': {
		return check_keyword(1, 3, "ive", TOKEN_GIVE);
	}
	case 'i':
		return check_keyword(1, 1, "f", TOKEN_IF);
	case 'l':
		return check_keyword(1, 2, "et", TOKEN_LET);
	case 'n':
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 'o': {
				return check_keyword(2, 1, "t", TOKEN_NOT);
			}
			case 'i': {
				return check_keyword(2, 1, "l", TOKEN_NIL);
			}
			case 'e': {
				return check_keyword(2, 1, "w", TOKEN_NEW);
			}
			default:;
			}
			break;
		}
		break;
	case 'o':
		return check_keyword(1, 1, "r", TOKEN_OR);
	case 'r':
		return check_keyword(1, 5, "eturn", TOKEN_RETURN);
	case 's': {
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 't': {
				if (scanner.current - scanner.start > 2) {
					return check_keyword(2, 4, "ruct",
							     TOKEN_STRUCT);
				}
			}
			default:;
			}
		}
		break;
	}
	case 'w':
		return check_keyword(1, 4, "hile", TOKEN_WHILE);
	case 'f': {
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 'a':
				return check_keyword(2, 3, "lse", TOKEN_FALSE);
			case 'o':
				return check_keyword(2, 1, "r", TOKEN_FOR);
			case 'n':
				return TOKEN_FN;
			case 'r':
				return check_keyword(2, 2, "om", TOKEN_FROM);
			default:;
			}
		}
		break;
	}
	case 'm': {
		return check_keyword(1, 4, "atch", TOKEN_MATCH);
	}
	case 't': {
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 'r':
				return check_keyword(2, 2, "ue", TOKEN_TRUE);
			case 'y':
				return check_keyword(2, 4, "peof",
						     TOKEN_TYPEOF);
			default:;
			}
		}
		break;
	}
	case 'u':
		return check_keyword(1, 2, "se", TOKEN_USE);
	case 'p':
		if (scanner.current - scanner.start > 1) {
			switch (scanner.start[1]) {
			case 'a':
				return check_keyword(2, 3, "nic", TOKEN_PANIC);
			case 'u':
				return check_keyword(2, 1, "b", TOKEN_PUB);
			default:;
			}
		}
		break;
	case 'E':
		return check_keyword(1, 2, "rr", TOKEN_ERR);
	case 'O':
		return check_keyword(1, 1, "k", TOKEN_OK);
	default:;
	}

	return TOKEN_IDENTIFIER;
}

void init_scanner(const char *source)
{
	scanner.start = source;
	scanner.current = source;
	scanner.line += 1;
}

/**
 * Creates a token of the given type from the current lexeme.
 * @param type The token type
 * @return The created token
 */
static Token make_token(const CruxTokenType type)
{
	Token token;
	token.type = type;
	token.start = scanner.start;
	token.length = (int)(scanner.current - scanner.start);
	token.line = scanner.line;
	return token;
}

/**
 * Creates an error token with the given message.
 * @param message The error message
 * @return The error token
 */
static Token error_token(const char *message)
{
	Token token;
	token.type = TOKEN_ERROR;
	token.start = message;
	token.length = (int)strlen(message);
	token.line = scanner.line;
	return token;
}

/**
 * Checks if the current character matches the expected one and advances if it
 * does.
 * @param expected The character to match
 * @return true if the character matches, false otherwise
 */
static bool match(const char expected)
{
	if (is_at_end())
		return false;
	if (*scanner.current != expected)
		return false;
	scanner.current++;
	return true;
}

/**
 * Scans a single-quoted string literal.
 * @return The string token or an error token
 */
static Token single_string(void)
{
	while (!is_at_end()) {
		if (peek() == '\'')
			break;
		if (peek() == '\\') {
			advance();
			if (!is_at_end()) {
				advance();
			}
			continue;
		}
		if (peek() == '\n') {
			scanner.line++;
		}
		advance();
	}

	if (is_at_end())
		return error_token("Unterminated String");
	// The closing quote
	advance();
	return make_token(TOKEN_STRING);
}

/**
 * Scans a double-quoted string literal.
 * @return The string token or an error token
 */
static Token double_string(void)
{
	while (!is_at_end()) {
		if (peek() == '"')
			break;
		if (peek() == '\\') {
			advance();
			if (!is_at_end()) {
				advance();
			}
			continue;
		}
		if (peek() == '\n') {
			scanner.line++;
		}
		advance();
	}

	if (is_at_end())
		return error_token("Unterminated String");
	// The closing quote
	advance();
	return make_token(TOKEN_STRING);
}

/**
 * Checks if a character is a digit (0-9).
 * @param c The character to check
 * @return true if the character is a digit, false otherwise
 */
static bool is_digit(const char c)
{
	return c >= '0' && c <= '9';
}

/**
 * Scans a numeric literal (integer or float).
 * @return The numeric token (TOKEN_INT or TOKEN_FLOAT)
 */
static Token number(void)
{
	while (is_digit(peek()))
		advance();
	bool fpFound = false;
	if (peek() == '.' && is_digit(peek_next())) {
		fpFound = true;
		advance();
		while (is_digit(peek()))
			advance();
	}
	if (fpFound) {
		return make_token(TOKEN_FLOAT);
	}
	return make_token(TOKEN_INT);
}

/**
 * Scans an identifier or keyword.
 * @return The identifier or keyword token
 */
static Token identifier(void)
{
	while (is_alpha(peek()) || is_digit(peek()))
		advance();
	return make_token(identifier_type());
}

Token scan_token(void)
{
	skip_whitespace();
	scanner.start = scanner.current;
	if (is_at_end())
		return make_token(TOKEN_EOF);

	const char c = advance();

	if (is_digit(c))
		return number();
	if (is_identifier_starter(c))
		return identifier();

	switch (c) {
	case ':':
		return make_token(TOKEN_COLON);
	case '(':
		return make_token(TOKEN_LEFT_PAREN);
	case ')':
		return make_token(TOKEN_RIGHT_PAREN);
	case '{':
		return make_token(TOKEN_LEFT_BRACE);
	case '}':
		return make_token(TOKEN_RIGHT_BRACE);
	case '[':
		return make_token(TOKEN_LEFT_SQUARE);
	case ']':
		return make_token(TOKEN_RIGHT_SQUARE);
	case ';':
		return make_token(TOKEN_SEMICOLON);
	case ',':
		return make_token(TOKEN_COMMA);
	case '.':
		return make_token(TOKEN_DOT);
	case '-':
		return make_token(match('=') ? TOKEN_MINUS_EQUAL : TOKEN_MINUS);
	case '+':
		return make_token(match('=') ? TOKEN_PLUS_EQUAL : TOKEN_PLUS);
	case '/':
		return make_token(match('=') ? TOKEN_SLASH_EQUAL : TOKEN_SLASH);
	case '\\':
		if (match('=')) {
			return make_token(TOKEN_BACK_SLASH_EQUAL);
		}
		return make_token(TOKEN_BACKSLASH);
	case '*': {
		if (match('*')) {
			return make_token(TOKEN_STAR_STAR);
		}
		if (match('=')) {
			return make_token(TOKEN_STAR_EQUAL);
		}
		return make_token(TOKEN_STAR);
	}
	case '%':
		if (match('=')) {
			return make_token(TOKEN_PERCENT_EQUAL);
		}
		return make_token(TOKEN_PERCENT);
	case '!': {
		if (match('=')) {
			return make_token(TOKEN_BANG_EQUAL);
		}
		return error_token("Unexpected token");
	}
	case '=': {
		if (match('=')) {
			return make_token(TOKEN_EQUAL_EQUAL);
		}
		if (match('>')) {
			return make_token(TOKEN_EQUAL_ARROW);
		}
		return make_token(TOKEN_EQUAL);
	}
	case '<':
		if (match('<')) {
			return make_token(TOKEN_LEFT_SHIFT);
		}
		return make_token(match('=') ? TOKEN_LESS_EQUAL : TOKEN_LESS);
	case '>':
		if (match('>')) {
			return make_token(TOKEN_RIGHT_SHIFT);
		}
		return make_token(match('=') ? TOKEN_GREATER_EQUAL
					     : TOKEN_GREATER);
	case '"':
		return double_string();
	case '\'':
		return single_string();
	case '?': {
		return make_token(TOKEN_QUESTION_MARK);
	}
	default:;
	}
	return error_token("Unexpected character.");
}

File: analysis/scanner.go
package analysis

type TokenType int

const (
	TOKEN_LEFT_PAREN TokenType = iota
	TOKEN_RIGHT_PAREN
	TOKEN_LEFT_BRACE
	TOKEN_RIGHT_BRACE
	TOKEN_LEFT_SQUARE
	TOKEN_RIGHT_SQUARE
	TOKEN_COMMA
	TOKEN_DOT
	TOKEN_MINUS
	TOKEN_PLUS
	TOKEN_SEMICOLON
	TOKEN_SLASH
	TOKEN_BACKSLASH
	TOKEN_STAR
	TOKEN_STAR_STAR
	TOKEN_PERCENT
	TOKEN_COLON
	TOKEN_SINGLE_QUOTE
	TOKEN_DOUBLE_QUOTE
	TOKEN_BANG_EQUAL
	TOKEN_EQUAL
	TOKEN_EQUAL_EQUAL
	TOKEN_GREATER
	TOKEN_GREATER_EQUAL
	TOKEN_LESS
	TOKEN_LESS_EQUAL
	TOKEN_LEFT_SHIFT
	TOKEN_RIGHT_SHIFT
	TOKEN_PLUS_EQUAL
	TOKEN_MINUS_EQUAL
	TOKEN_STAR_EQUAL
	TOKEN_SLASH_EQUAL
	TOKEN_BACK_SLASH_EQUAL
	TOKEN_PERCENT_EQUAL
	TOKEN_IDENTIFIER
	TOKEN_STRING
	TOKEN_INT
	TOKEN_FLOAT
	TOKEN_AND
	TOKEN_NOT
	TOKEN_ELSE
	TOKEN_FALSE
	TOKEN_FOR
	TOKEN_FN
	TOKEN_IF
	TOKEN_NIL
	TOKEN_OR
	TOKEN_RETURN
	TOKEN_TRUE
	TOKEN_LET
	TOKEN_WHILE
	TOKEN_ERROR
	TOKEN_BREAK
	TOKEN_CONTINUE
	TOKEN_USE
	TOKEN_FROM
	TOKEN_PUB
	TOKEN_AS
	TOKEN_EOF
	TOKEN_MATCH
	TOKEN_EQUAL_ARROW
	TOKEN_OK
	TOKEN_ERR
	TOKEN_DEFAULT
	TOKEN_GIVE
	TOKEN_TYPEOF
	TOKEN_NEW
	TOKEN_PANIC
	TOKEN_DOLLAR_LEFT_CURLY
	TOKEN_DOLLAR_LEFT_SQUARE
	TOKEN_STRUCT
	TOKEN_QUESTION_MARK
)

var keywords = map[string]TokenType{
	"and":      TOKEN_AND,
	"not":      TOKEN_NOT,
	"else":     TOKEN_ELSE,
	"false":    TOKEN_FALSE,
	"for":      TOKEN_FOR,
	"fn":       TOKEN_FN,
	"if":       TOKEN_IF,
	"nil":      TOKEN_NIL,
	"or":       TOKEN_OR,
	"return":   TOKEN_RETURN,
	"true":     TOKEN_TRUE,
	"let":      TOKEN_LET,
	"while":    TOKEN_WHILE,
	"break":    TOKEN_BREAK,
	"continue": TOKEN_CONTINUE,
	"use":      TOKEN_USE,
	"from":     TOKEN_FROM,
	"pub":      TOKEN_PUB,
	"as":       TOKEN_AS,
	"match":    TOKEN_MATCH,
	"Ok":       TOKEN_OK,
	"Err":      TOKEN_ERR,
	"default":  TOKEN_DEFAULT,
	"give":     TOKEN_GIVE,
	"typeof":   TOKEN_TYPEOF,
	"new":      TOKEN_NEW,
	"panic":    TOKEN_PANIC,
	"struct":   TOKEN_STRUCT,
}

type Token struct {
	Type    TokenType
	Literal string
	Line    int
	Column  int
}

type Scanner struct {
	source  string
	start   int
	current int
	line    int
	tokens  []Token
}

func NewScanner(source string) *Scanner {
	return &Scanner{
		source:  source,
		start:   0,
		current: 0,
		line:    1,
		tokens:  []Token{},
	}
}

func (s *Scanner) isAtEnd() bool {
	return s.current >= len(s.source)
}

func (s *Scanner) advance() byte {
	c := s.source[s.current]
	s.current++
	return c
}

func (s *Scanner) peek() byte {
	if s.isAtEnd() {
		return 0
	}
	return s.source[s.current]
}

func (s *Scanner) peekNext() byte {
	if s.current+1 >= len(s.source) {
		return 0
	}
	return s.source[s.current+1]
}

func (s *Scanner) match(expected byte) bool {
	if s.isAtEnd() {
		return false
	}
	if s.source[s.current] != expected {
		return false
	}
	s.current++
	return true
}

func (s *Scanner) isAlpha(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

func (s *Scanner) isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

func (s *Scanner) isIdentifierStart(c byte) bool {
	return s.isAlpha(c) || c == '_'
}

func (s *Scanner) isIdentifier(c byte) bool {
	return s.isIdentifierStart(c) || s.isDigit(c)
}

func (s *Scanner) skipWhitespace() {
	for !s.isAtEnd() {
		c := s.peek()
		switch c {
		case ' ', '\r', '\t':
			s.advance()
		case '\n':
			s.line++
			s.advance()
		case '/':
			if s.peekNext() == '/' {
				for s.peek() != '\n' && !s.isAtEnd() {
					s.advance()
				}
			} else {
				return
			}
		default:
			return
		}
	}
}

func (s *Scanner) scanString(quote byte) Token {
	start := s.line
	for s.peek() != quote && !s.isAtEnd() {
		if s.peek() == '\n' {
			s.line++
		}
		if s.peek() == '\\' && !s.isAtEnd() {
			s.advance()
			if !s.isAtEnd() {
				s.advance()
			}
			continue
		}
		s.advance()
	}

	if s.isAtEnd() {
		return Token{Type: TOKEN_ERROR, Literal: "Unterminated String", Line: start}
	}

	s.advance()
	return Token{Type: TOKEN_STRING, Literal: s.source[s.start+1 : s.current-1], Line: start}
}

func (s *Scanner) scanNumber() Token {
	start := s.line
	hasDecimal := false

	for s.isDigit(s.peek()) {
		s.advance()
	}

	if s.peek() == '.' && s.isDigit(s.peekNext()) {
		hasDecimal = true
		s.advance()
		for s.isDigit(s.peek()) {
			s.advance()
		}
	}

	tokenType := TOKEN_INT
	if hasDecimal {
		tokenType = TOKEN_FLOAT
	}

	return Token{Type: tokenType, Literal: s.source[s.start:s.current], Line: start}
}

func (s *Scanner) scanIdentifier() Token {
	for s.isIdentifier(s.peek()) {
		s.advance()
	}

	text := s.source[s.start:s.current]
	tokenType, ok := keywords[text]
	if !ok {
		tokenType = TOKEN_IDENTIFIER
	}

	return Token{Type: tokenType, Literal: text, Line: s.line}
}

func (s *Scanner) makeToken(tokenType TokenType) Token {
	return Token{
		Type:    tokenType,
		Literal: s.source[s.start:s.current],
		Line:    s.line,
		Column:  s.start,
	}
}

func (s *Scanner) errorToken(message string) Token {
	return Token{
		Type:    TOKEN_ERROR,
		Literal: message,
		Line:    s.line,
		Column:  s.start,
	}
}

func (s *Scanner) ScanToken() Token {
	s.skipWhitespace()
	s.start = s.current

	if s.isAtEnd() {
		return s.makeToken(TOKEN_EOF)
	}

	c := s.advance()

	if s.isDigit(c) {
		return s.scanNumber()
	}

	if s.isIdentifierStart(c) {
		return s.scanIdentifier()
	}

	switch c {
	case '(':
		return s.makeToken(TOKEN_LEFT_PAREN)
	case ')':
		return s.makeToken(TOKEN_RIGHT_PAREN)
	case '{':
		return s.makeToken(TOKEN_LEFT_BRACE)
	case '}':
		return s.makeToken(TOKEN_RIGHT_BRACE)
	case '[':
		return s.makeToken(TOKEN_LEFT_SQUARE)
	case ']':
		return s.makeToken(TOKEN_RIGHT_SQUARE)
	case ';':
		return s.makeToken(TOKEN_SEMICOLON)
	case ',':
		return s.makeToken(TOKEN_COMMA)
	case '.':
		return s.makeToken(TOKEN_DOT)
	case '-':
		if s.match('=') {
			return s.makeToken(TOKEN_MINUS_EQUAL)
		}
		return s.makeToken(TOKEN_MINUS)
	case '+':
		if s.match('=') {
			return s.makeToken(TOKEN_PLUS_EQUAL)
		}
		return s.makeToken(TOKEN_PLUS)
	case '/':
		if s.match('=') {
			return s.makeToken(TOKEN_SLASH_EQUAL)
		}
		return s.makeToken(TOKEN_SLASH)
	case '\\':
		if s.match('=') {
			return s.makeToken(TOKEN_BACK_SLASH_EQUAL)
		}
		return s.makeToken(TOKEN_BACKSLASH)
	case '*':
		if s.match('*') {
			return s.makeToken(TOKEN_STAR_STAR)
		}
		if s.match('=') {
			return s.makeToken(TOKEN_STAR_EQUAL)
		}
		return s.makeToken(TOKEN_STAR)
	case '%':
		if s.match('=') {
			return s.makeToken(TOKEN_PERCENT_EQUAL)
		}
		return s.makeToken(TOKEN_PERCENT)
	case '!':
		if s.match('=') {
			return s.makeToken(TOKEN_BANG_EQUAL)
		}
		return s.errorToken("Unexpected token")
	case '=':
		if s.match('=') {
			return s.makeToken(TOKEN_EQUAL_EQUAL)
		}
		if s.match('>') {
			return s.makeToken(TOKEN_EQUAL_ARROW)
		}
		return s.makeToken(TOKEN_EQUAL)
	case '<':
		if s.match('<') {
			return s.makeToken(TOKEN_LEFT_SHIFT)
		}
		if s.match('=') {
			return s.makeToken(TOKEN_LESS_EQUAL)
		}
		return s.makeToken(TOKEN_LESS)
	case '>':
		if s.match('>') {
			return s.makeToken(TOKEN_RIGHT_SHIFT)
		}
		if s.match('=') {
			return s.makeToken(TOKEN_GREATER_EQUAL)
		}
		return s.makeToken(TOKEN_GREATER)
	case '"':
		return s.scanString('"')
	case '\'':
		return s.scanString('\'')
	case '?':
		return s.makeToken(TOKEN_QUESTION_MARK)
	case ':':
		return s.makeToken(TOKEN_COLON)
	}

	return s.errorToken("Unexpected character.")
}

func (s *Scanner) ScanAll() []Token {
	s.line = 1
	s.start = 0
	s.current = 0

	tokens := []Token{}
	for {
		token := s.ScanToken()
		tokens = append(tokens, token)
		if token.Type == TOKEN_EOF || token.Type == TOKEN_ERROR {
			break
		}
	}
	return tokens
}
